src/ray/raylet/node_manager.cc

```diff
@@ -1313,15 +1321,25 @@ void NodeManager::AssignTask(Task &task) {
           task_dependency_manager_.UnsubscribeDependencies(spec.TaskId());
         } else {
           RAY_LOG(WARNING) << "Failed to send task to worker, disconnecting client";
+          // Queue this task for future assignment. We need to do this since
+          // DispatchTasks() removed it from the ready queue. The task will be
+          // assigned to a worker once one becomes available.
+          // (See design_docs/task_states.rst for the state transition diagram.)
+          if (!local_queues_.HasTask(task.GetTaskSpecification().TaskId())) {
+            // Make sure we don't insert a duplicate task; this could happen
+            // if AssignTask() is recursivelly called on same task.
+            local_queues_.QueueReadyTasks(std::vector<Task>({task}));
```

Comment:
This if statement is a little awkward; we should just make sure that at this point, the task is not in any of the queues. I think we can do this by either catching the return value of `AssignTask` below and only re-queuing the task if the task was not assigned, or changing it to a `DispatchTasks({task})`.

Also, I'm realizing we probably want the latter since currently we do the check for whether there are enough resources for the task to run in `DispatchTasks` (although it's debatable whether that check should go in `AssignTask` instead).